Handle callback errors

lo1tuma opened the pull request:
While debugging a failing test I’ve noticed that there was an unhandled callback error in the failing test. I spent a lot of time to figure out why the assertion was failing, but the data, used in the assertion, was not present at all due to a previous error.

I’ve added eslint and enabled to rule [`handle-callback-err`](http://eslint.org/docs/rules/handle-callback-err) which can detect some unhandled errors statically. ESLint found round about 100 unhandled errors in the whole project.

I tried to fix this and handle all errors properly. In some cases I was not sure if an error was ignored intentionally. So please review these changes carefully and let me know if there is such an case where the error should be ignored. Nevertheless, even if an error should be ignored I think we should at least log the error with winston.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NodeBB/NodeBB?pullRequest=4951) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NodeBB/NodeBB?pullRequest=4951) before we can accept your contribution.<br/>

lo1tuma commented:
The linting process is failing on travis for node versions < 4. ESLint requires at least node version 4, as it uses some ES2015 features.

barisusakli commented:
Can you make the eslint addition a separate PR so we can merge the callback additions?

Also you can just use `winston.error(err);`

barisusakli commented on app.js:
```diff
@@ -296,9 +301,19 @@ function activate() {
 		var db = require('./src/database');
 
 		db.getSortedSetRange('plugins:active', 0, -1, function(err, plugins) {
+		if (err) {
```
indentation is off here

lo1tuma replied:
fixed

barisusakli commented on install/web.js:
```diff
@@ -130,6 +130,10 @@ function compileLess(callback) {
 	fs.readFile(path.join(__dirname, '../public/less/install.less'), function(err, style) {
+		if (err) {
+			return winston.error('Unable to read LESS install file: ', err);
+		};
```
extra `;` here

lo1tuma replied:
yikes 🙈 
should be fixed now

barisusakli commented on src/groups/cover.js:
```diff
@@ -70,6 +70,10 @@ module.exports = function(Groups) {
 			if (err) {
 				return fs.unlink(tempPath, function(unlinkErr) {
+					if (unlinkErr) {
+						return callback(unlinkErr);
```
Can you just `winston.error(err)` this and send back the original error.

lo1tuma replied:
done

barisusakli commented on src/languages.js:
```diff
@@ -30,6 +30,10 @@ Languages.get = function(code, key, callback) {
 	var languageData;
 
 	fs.readFile(path.join(__dirname, '../public/language/', code, key), { encoding: 'utf-8' }, function(err, data) {
+		if (err) {
```
remove this, the below code handles the missing files.

lo1tuma replied:
How do you know that the error always indicates that a file is missing? Also if an error exists I would assume that messageData is null or undefined, so accessing the parsedMessage property would fail.

barisusakli replied:
Yeah the correct thing to do would be to check err.code like `e.code === 'ENOENT'`

lo1tuma replied:
Good idea.

I’ve added an additional check to make sure the callback is only called with `err` when `err.code !== 'ENOENT'`.

barisusakli commented on src/messaging.js:
```diff
@@ -209,6 +209,10 @@ var async = require('async'),
 			};
 
 			plugins.fireHook('filter:messaging.parse', messageData, function(err, messageData) {
+				if (err) {
```
remove this, I will have to fix this function first before adding err, looks like the function is sending the result in the first param as well.

lo1tuma replied:
How do you know that the error always indicates that a file is missing? Also if an error exists I would assume that `messageData` is `null` or `undefined`, so accessing the `parsedMessage` property would fail.

lo1tuma replied:
done

barisusakli commented on src/user/picture.js:
```diff
@@ -200,6 +200,10 @@ module.exports = function(User) {
 			if (err) {
 				return fs.unlink(data.file.path, function(unlinkErr) {
+					if (unlinkErr) {
+						return callback(unlinkErr);
```
just log this errors with winstin.error()

lo1tuma replied:
done

lo1tuma commented:
@barisusakli Thanks for the code review. I addressed most of the concerns you mentioned, except for one, where I posted a follow-up question.

Apart from that, I also removed the ESLint part from this PR. I will create a separate PR for this :wink.
